File: manual/java-src/com/atta/mcpanel/overlay/tabs/ServerTab.java
package com.atta.mcpanel.overlay.tabs;

import android.content.Context;
import android.text.InputType;
import android.view.Gravity;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.atta.mcpanel.core.Palette;
import com.atta.mcpanel.overlay.PanelHost;
import com.atta.mcpanel.overlay.UiKit;

/** زبانهٔ «سرور» — اتصال RCON و کنسول سرورهای شخصی */
public final class ServerTab {

    private ServerTab() {}

    private static final int LOG_LIMIT = 90000;

    public static View build(final PanelHost host) {
        final Context ctx = host.getContext();
        final LinearLayout col = UiKit.vcol(ctx, 14);

        // ---- اتصال
        col.addView(UiKit.sectionLabel(ctx, "اتصال RCON به سرور شخصی"));
        col.addView(UiKit.caption(ctx,
                "سرور روی همین دستگاه: localhost · روی سیستم دیگر: IP همان سیستم در شبکه", false));

        final EditText hostEt = host.newInput(
                "آدرس سرور", InputType.TYPE_CLASS_TEXT, host.getPrefs().getRconHost());
        final EditText portEt = host.newInput(
                "پورت", InputType.TYPE_CLASS_NUMBER,
                String.valueOf(host.getPrefs().getRconPort()));
        final EditText passEt = host.newInput(
                "رمز RCON", InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD,
                host.getPrefs().getRconPassword());

        LinearLayout topRow = UiKit.hrow(ctx);
        topRow.addView(hostEt, new LinearLayout.LayoutParams(
                0, UiKit.dp(ctx, 46), 1f));
        topRow.addView(portEt, new LinearLayout.LayoutParams(
                UiKit.dp(ctx, 92), UiKit.dp(ctx, 46)));
        col.addView(topRow, UiKit.wrapParams(topRow, 4, 46));
        col.addView(passEt, UiKit.wrapParams(passEt, 4, 46));

        // ---- نوار وضعیت
        final LinearLayout statusRow = UiKit.hrow(ctx);
        statusRow.setPadding(0, UiKit.dp(ctx, 8), 0, UiKit.dp(ctx, 4));
        final View dot = UiKit.dot(ctx, Palette.DOT_OFF, 10);
        statusRow.addView(dot, new LinearLayout.LayoutParams(
                UiKit.dp(ctx, 10), UiKit.dp(ctx, 10)));
        ((LinearLayout.LayoutParams) dot.getLayoutParams()).leftMargin = UiKit.dp(ctx, 8);
        ((LinearLayout.LayoutParams) dot.getLayoutParams()).rightMargin = UiKit.dp(ctx, 2);
        final TextView tvState = new TextView(ctx);
        tvState.setTextSize(12.5f);
        tvState.setTextColor(Palette.TEXT_SUB);
        statusRow.addView(tvState, new LinearLayout.LayoutParams(
                0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f));
        final TextView btnConnect = UiKit.chip(ctx, "اتصال", UiKit.KIND_ACCENT, null);
        statusRow.addView(btnConnect, new LinearLayout.LayoutParams(
                UiKit.dp(ctx, 120), UiKit.dp(ctx, 42)));
        col.addView(statusRow);

        // ---- کنسول
        col.addView(UiKit.sectionLabel(ctx, "کنسول سرور"));
        final ScrollView logScroll = new ScrollView(ctx);
        logScroll.setBackgroundColor(Palette.LOG_BG);
        logScroll.setVerticalScrollBarEnabled(false);
        logScroll.setOverScrollMode(View.OVER_SCROLL_NEVER);
        final TextView logTv = new TextView(ctx);
        logTv.setTextSize(12f);
        logTv.setTextColor(Palette.TEXT_MAIN);
        logTv.setLineSpacing(0f, 1.1f);
        logTv.setPadding(UiKit.dp(ctx, 10), UiKit.dp(ctx, 8),
                UiKit.dp(ctx, 10), UiKit.dp(ctx, 8));
        logScroll.addView(logTv, new ScrollView.LayoutParams(
                ScrollView.LayoutParams.MATCH_PARENT,
                ScrollView.LayoutParams.WRAP_CONTENT));
        LinearLayout.LayoutParams logLp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, UiKit.dp(ctx, 170));
        logLp.setMargins(UiKit.dp(ctx, 4), 0, UiKit.dp(ctx, 4), 0);
        col.addView(logScroll, logLp);

        host.registerConsoleListener(new PanelHost.ConsoleListener() {
            @Override public void onLine(String line) {
                if (logTv.length() > 0) logTv.append("\n");
                logTv.append(line);
                if (logTv.length() > LOG_LIMIT) {
                    logTv.setText(logTv.getText().subSequence(
                            logTv.length() - LOG_LIMIT, logTv.length()));
                }
                logScroll.post(new Runnable() {
                    @Override public void run() {
                        logScroll.fullScroll(View.FOCUS_DOWN);
                    }
                });
            }
        });

        // ---- ارسال دستور
        final LinearLayout sendRow = UiKit.hrow(ctx);
        final EditText inputEt = host.newInput(
                "دستور اپراتور — بدون / مثل op Steve", InputType.TYPE_CLASS_TEXT);
        sendRow.addView(inputEt, new LinearLayout.LayoutParams(
                0, UiKit.dp(ctx, 46), 1f));
        final TextView btnSend = UiKit.chip(ctx, "ارسال", UiKit.KIND_ACCENT, new Runnable() {
            @Override public void run() {
                String cmd = inputEt.getText().toString().trim();
                if (cmd.isEmpty()) return;
                logTv.append("\n> " + cmd);
                logScroll.post(new Runnable() {
                    @Override public void run() { logScroll.fullScroll(View.FOCUS_DOWN); }
                });
                host.sendConsole(cmd);
                inputEt.setText("");
            }
        });
        sendRow.addView(btnSend, new LinearLayout.LayoutParams(
                UiKit.dp(ctx, 84), UiKit.dp(ctx, 46)));
        col.addView(sendRow, UiKit.wrapParams(sendRow, 4, 46));

        // ---- دستورهای سریع
        col.addView(UiKit.sectionLabel(ctx, "دستورهای سریع سرور", false));
        UiKit.chipRow(ctx, new TextView[]{
                UiKit.chip(ctx, "👥 فهرست بازیکن‌ها", new Runnable() {
                    @Override public void run() { host.sendConsole("list"); }
                }),
                UiKit.chip(ctx, "🌱 بذر دنیا", new Runnable() {
                    @Override public void run() { host.sendConsole("seed"); }
                }),
                UiKit.chip(ctx, "💾 ذخیرهٔ فوری", new Runnable() {
                    @Override public void run() { host.sendConsole("save-all"); }
                }),
                UiKit.chip(ctx, "❓ help", new Runnable() {
                    @Override public void run() { host.sendConsole("help"); }
                }),
        }, host.cols(4), col);

        final TextView btnStop = UiKit.chip(ctx, "⛔ استاپ کامل سرور", UiKit.KIND_DANGER,
                new Runnable() {
                    @Override public void run() {
                        host.confirm("استاپ سرور",
                                "سرور برای همهٔ بازیکن‌ها بسته می‌شود. مطمئنی؟",
                                new Runnable() {
                                    @Override public void run() { host.sendConsole("stop"); }
                                });
                    }
                });
        LinearLayout.LayoutParams stopLp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, UiKit.dp(ctx, 46));
        stopLp.setMargins(UiKit.dp(ctx, 4), UiKit.dp(ctx, 6), UiKit.dp(ctx, 4), 0);
        col.addView(btnStop, stopLp);

        // ---- به‌روزرسانی وضعیت
        host.registerStateListener(new PanelHost.StateListener() {
            @Override public void onState() { updateState(host, dot, tvState, btnConnect); }
        });
        updateState(host, dot, tvState, btnConnect);

        btnConnect.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) {
                if (host.isRconConnected()) {
                    host.rconDisconnect();
                } else {
                    saveFields(host, hostEt, portEt, passEt);
                    if (host.rconConnect()) {
                        logTv.append("\n→ در حال اتصال به " + host.getPrefs().getRconHost()
                                + ":" + host.getPrefs().getRconPort() + " …");
                        logScroll.post(new Runnable() {
                            @Override public void run() {
                                logScroll.fullScroll(View.FOCUS_DOWN);
                            }
                        });
                    }
                }
            }
        });
        btnConnect.setClickable(true);

        ScrollView sv = new ScrollView(ctx);
        sv.setFillViewport(true);
        sv.setOverScrollMode(View.OVER_SCROLL_NEVER);
        sv.addView(col);
        return sv;
    }

    private static void saveFields(PanelHost host, EditText h, EditText p, EditText pass) {
        host.getPrefs().setRconHost(h.getText().toString());
        String pt = p.getText().toString().trim();
        host.getPrefs().setRconPort(pt.isEmpty() ? host.getPrefs().getRconPort()
                : Integer.parseInt(pt));
        host.getPrefs().setRconPassword(pass.getText().toString());
    }

    private static void updateState(PanelHost host, View dot, TextView tv, TextView btn) {
        boolean connected = host.isRconConnected();
        boolean attempting = host.isRconAttempting();
        int color = connected ? Palette.DOT_OK : (attempting ? Palette.DOT_BUSY : Palette.DOT_BAD);
        UiKit.dotColor(dot, color);
        tv.setText(connected ? "وصل است — می‌توانی دستور بدهی"
                : attempting ? "در حال اتصال…" : "آفلاین — دکمهٔ اتصال را بزن");
        tv.setTextColor(connected ? Palette.DOT_OK : Palette.TEXT_SUB);
        btn.setText(connected ? "قطع اتصال" : "اتصال");
        btn.setGravity(Gravity.CENTER);
    }
}

File: app/src/main/java/com/atta/mcpanel/overlay/tabs/ServerTab.kt
package com.atta.mcpanel.overlay.tabs

import android.text.InputType
import android.view.Gravity
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.atta.mcpanel.R
import com.atta.mcpanel.overlay.PanelHost
import com.atta.mcpanel.overlay.UiKit
import com.google.android.material.button.MaterialButton

/**
 * زبانهٔ «سرور»: اتصال RCON به سرورهای شخصی (هم روی همین دستگاه، هم شبکهٔ داخلی)
 * و کنسول کامل دستورهای اپراتور.
 */
object ServerTab {

    private const val LOG_LIMIT = 120000

    fun build(host: PanelHost): View {
        val ctx = host.context
        val col = LinearLayout(ctx).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(UiKit.dp(ctx, 14), UiKit.dp(ctx, 6), UiKit.dp(ctx, 14), UiKit.dp(ctx, 8))
        }

        // ---- اتصال --------------------------------------------------------
        col.addView(UiKit.sectionLabel(ctx, "اتصال RCON به سرور شخصی"))
        col.addView(
            UiKit.caption(
                ctx,
                "آدرس، پورت و رمز را بگذار؛ اگر سرور روی همین گوشی/کامپیوتر است: localhost و اگر روی سیستم دیگر است: IP همان سیستم."
            )
        )

        val hostEt = host.newInput("آدرس سرور", InputType.TYPE_CLASS_TEXT, host.prefs.rconHost)
        val portEt = host.newInput("پورت", InputType.TYPE_CLASS_NUMBER, host.prefs.rconPort.toString())
        val passEt = host.newInput(
            "رمز RCON",
            InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD,
            host.prefs.rconPassword,
        )

        val topRow = LinearLayout(ctx).apply { orientation = LinearLayout.HORIZONTAL }
        topRow.addView(
            hostEt,
            LinearLayout.LayoutParams(0, UiKit.dp(ctx, 46), 1f).apply {
                setMargins(UiKit.dp(ctx, 4), UiKit.dp(ctx, 4), UiKit.dp(ctx, 2), 0)
            }
        )
        topRow.addView(
            portEt,
            LinearLayout.LayoutParams(UiKit.dp(ctx, 96), UiKit.dp(ctx, 46)).apply {
                setMargins(UiKit.dp(ctx, 2), UiKit.dp(ctx, 4), UiKit.dp(ctx, 4), 0)
            }
        )
        col.addView(topRow)
        col.addView(
            passEt,
            LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, UiKit.dp(ctx, 46)
            ).apply {
                setMargins(UiKit.dp(ctx, 4), UiKit.dp(ctx, 4), UiKit.dp(ctx, 4), 0)
            }
        )

        fun saveFields() {
            host.prefs.rconHost = hostEt.text.toString()
            host.prefs.rconPort = portEt.text.toString().toIntOrNull() ?: host.prefs.rconPort
            host.prefs.rconPassword = passEt.text.toString()
        }

        // نوار وضعیت + دکمهٔ اتصال
        val statusRow = LinearLayout(ctx).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            setPadding(UiKit.dp(ctx, 4), UiKit.dp(ctx, 10), UiKit.dp(ctx, 4), UiKit.dp(ctx, 2))
        }
        val dot = UiKit.dot(ctx, ContextCompat.getColor(ctx, R.color.dotOff))
        statusRow.addView(
            dot,
            LinearLayout.LayoutParams(UiKit.dp(ctx, 10), UiKit.dp(ctx, 10)).apply {
                setMargins(0, 0, UiKit.dp(ctx, 8), 0)
            }
        )
        val tvState = TextView(ctx).apply {
            textSize = 13f
            setTextColor(ContextCompat.getColor(ctx, R.color.textSub))
        }
        statusRow.addView(tvState, LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f))
        val btnConnect = MaterialButton(ctx).apply {
            text = "اتصال"
            isAllCaps = false
            textSize = 13f
            minHeight = 0
            insetTop = 0
            insetBottom = 0
        }
        btnConnect.layoutParams = LinearLayout.LayoutParams(UiKit.dp(ctx, 118), UiKit.dp(ctx, 42))
        statusRow.addView(btnConnect)
        col.addView(statusRow)

        // ---- کنسول --------------------------------------------------------
        col.addView(UiKit.sectionLabel(ctx, "کنسول سرور", accent = true))

        val logScroll = ScrollView(ctx).apply {
            setBackgroundResource(R.drawable.bg_log)
            isVerticalScrollBarEnabled = false
            isFillViewport = false
            overScrollMode = View.OVER_SCROLL_NEVER
        }
        val logTv = TextView(ctx).apply {
            textSize = 12f
            setTextColor(ContextCompat.getColor(ctx, R.color.textMain))
            setLineSpacing(0f, 1.15f)
            setPadding(UiKit.dp(ctx, 10), UiKit.dp(ctx, 8), UiKit.dp(ctx, 10), UiKit.dp(ctx, 8))
        }
        logScroll.addView(
            logTv,
            ScrollView.LayoutParams(
                ScrollView.LayoutParams.MATCH_PARENT, ScrollView.LayoutParams.WRAP_CONTENT
            )
        )
        col.addView(
            logScroll,
            LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f
            ).apply {
                setMargins(UiKit.dp(ctx, 4), 0, UiKit.dp(ctx, 4), 0)
            }
        )

        fun append(line: String) {
            if (logTv.text.isNotEmpty()) logTv.append("\n")
            logTv.append(line)
            if (logTv.text.length > LOG_LIMIT) {
                logTv.text = logTv.text.substring(logTv.text.length - LOG_LIMIT)
            }
            logScroll.post { logScroll.fullScroll(View.FOCUS_DOWN) }
        }

        host.registerConsoleListener { line -> append(line) }

        // ---- ارسال دستور --------------------------------------------------
        val sendRow = LinearLayout(ctx).apply { orientation = LinearLayout.HORIZONTAL }
        val inputEt: EditText = host.newInput(
            "دستور اپراتور — بدون / بنویس، مثلاً op Steve", InputType.TYPE_CLASS_TEXT
        )
        sendRow.addView(inputEt, LinearLayout.LayoutParams(0, UiKit.dp(ctx, 44), 1f))
        val btnSend = UiKit.chip(ctx, "ارسال", UiKit.ChipKind.ACCENT) {
            val cmd = inputEt.text.toString().trim()
            if (cmd.isNotEmpty()) {
                append("> $cmd")
                host.sendConsole(cmd)
            }
        }
        sendRow.addView(btnSend, LinearLayout.LayoutParams(UiKit.dp(ctx, 74), UiKit.dp(ctx, 44)))
        col.addView(sendRow)
        col.addView(UiKit.caption(ctx, "خروجی کنسول و پیام خطاها اینجا می‌آید.", dim = true))

        // ---- دستورهای پرکاربرد سرور -------------------------------------
        col.addView(UiKit.sectionLabel(ctx, "دستورهای سریع سرور", accent = false))
        UiKit.chipRow(
            ctx,
            listOf(
                UiKit.chip(ctx, "👥 فهرست بازیکن‌ها") { host.sendConsole("list") },
                UiKit.chip(ctx, "🌱 بذر دنیا") { host.sendConsole("seed") },
                UiKit.chip(ctx, "💾 ذخیرهٔ فوری") { host.sendConsole("save-all") },
                UiKit.chip(ctx, "❓ help") { host.sendConsole("help") },
            ),
            4, col,
        )

        val btnStop = UiKit.chip(ctx, "⛔ استاپ کامل سرور", UiKit.ChipKind.DANGER) {
            host.confirm(
                "استاپ سرور",
                "سرور برای همهٔ بازیکن‌ها بسته می‌شود و همه از بازی بیرون می‌افتند. مطمئنی؟",
            ) { host.sendConsole("stop") }
        }
        col.addView(
            btnStop,
            LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, UiKit.dp(ctx, 46)
            ).apply {
                val m = UiKit.dp(ctx, 4)
                setMargins(m, UiKit.dp(ctx, 8), m, 0)
            }
        )

        // ---- به‌روزرسانی وضعیت ------------------------------------------
        fun updateStateUi() {
            val connected = host.isRconConnected()
            val attempting = host.rconAttempting
            val color = ContextCompat.getColor(
                ctx,
                when {
                    connected -> R.color.dotOk
                    attempting -> R.color.dotBusy
                    else -> R.color.dotBad
                }
            )
            (dot.background.mutate() as android.graphics.drawable.GradientDrawable)
                .setColor(color)
            tvState.text = when {
                connected -> "وصل است — می‌توانی دستور بدهی"
                attempting -> "در حال اتصال…"
                else -> "آفلاین — وصل شو تا دستورها اجرا شوند"
            }
            tvState.setTextColor(
                ContextCompat.getColor(
                    ctx,
                    when {
                        connected -> R.color.dotOk
                        attempting -> R.color.dotBusy
                        else -> R.color.textSub
                    }
                )
            )
            btnConnect.text = if (connected) "قطع اتصال" else "اتصال"
        }

        btnConnect.setOnClickListener {
            if (host.isRconConnected()) {
                host.rconDisconnect()
            } else {
                saveFields()
                if (host.rconConnect()) {
                    updateStateUi()
                    append("→ در حال اتصال به ${host.prefs.rconHost}:${host.prefs.rconPort} …")
                }
            }
        }
        host.registerStateListener { updateStateUi() }
        updateStateUi()
        return col
    }
}
